# Credit Card Fraud Detection System

### (1) Install requiered packages(dependencies)

In [1]:
# installed dependencies 
!pip install numpy
!pip install pandas
!pip install scikit-learn
!pip install opendatasets
!pip install joblib

In [2]:
# import installed packages
import joblib
import numpy as np
import pandas as pd
import opendatasets as od
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


### (2) Get access from kaggle or upload the dataset

#### > Once run the cell below you can skip the cell for next time

In [10]:
# Specify the Kaggle dataset URL
kaggle_dataset_url = "https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud"
# Download the Kaggle dataset
od.download(kaggle_dataset_url)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username:

  zeaman44


Your Kaggle Key:

  ········


100%|███████████████████████████████████████████████████████████████████| 66.0M/66.0M [00:06<00:00, 10.2MB/s]


# START HERE

In [3]:
# Upload the dataset
creditCard_dataset = ("creditcardfraud/creditcard.csv")

In [4]:
# Upload the dataset to Pandas dataframe
creditCard_df = pd.read_csv(creditCard_dataset)

### (3) Visualize the dataset

In [5]:
# Print out the first five rows of the dataset
creditCard_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [6]:
# Print out the last five rows of the dataset
creditCard_df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [7]:
# Extracting & print out the general information of the dataframe
creditCard_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [24]:
# Interpreting:Each column contains numerical (float64) values, and there are no missing values (non-null count matches the total count of rows, 284,807).

In [8]:
# Visulize the Normal (Legit) transaction and fraud transaction
creditCard_df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

#### Interpreting:
#### -> Value "0" has a count of 284,315 occurrences.
#### -> Value "1" has a count of 492 occurrences
#### That indicated a highly biased dataset
#### Where:
#### 0 => Normal Transactions
#### 1 => Fraud Transactions

In [9]:
# Assigning Classes in Pandas DataFrame
normal = creditCard_df[creditCard_df.Class == 0]
print(normal.shape)
fraud = creditCard_df[creditCard_df.Class == 1]
print(fraud.shape)

(284315, 31)
(492, 31)


In [10]:
# Print out the statistical measures of the above dataframe
normal.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [12]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [13]:
# Print out the values for both transactions after the dataset is cleaned and unbiased
creditCard_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

492

In [14]:
normal_sample = normal.sample(n = fraud.shape[0])

In [15]:
creditCard_new_df = pd.concat([normal_sample, fraud], axis = 0)

In [16]:
# Visualized the new datast/dataframe
creditCard_new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
100818   67635.0  1.205203 -1.273859  1.688107 -0.269991 -2.002077  0.748282   
143495   85391.0 -0.438080  1.047487  1.698626  0.046575 -0.242688 -1.083626   
154204  100912.0  1.970590 -0.490162  0.070065  0.449218 -0.774299  0.054676   
141532   84378.0  1.116321 -0.521017  0.870835 -0.936391 -0.809979  0.433784   
123720   77036.0 -0.672101  0.605524  1.907994 -0.239642 -0.403889  0.266418   

              V7        V8        V9  ...       V21       V22       V23  \
100818 -1.856858  0.599615  0.721384  ...  0.211380  0.873052 -0.027948   
143495  0.714122 -0.084256 -0.408340  ... -0.206369 -0.540060  0.022949   
154204 -1.086064  0.102273  2.755917  ... -0.124383 -0.029048  0.369948   
141532 -0.854976  0.411624  1.605046  ...  0.091546  0.530595  0.016285   
123720  0.046336  0.269681 -0.073173  ...  0.220005  0.514393  0.019879   

             V24       V25       V26       V27       V28  Amount  Class  
100818  0.242419  0.284600 -0.039573  0.086433  0.011910    5.94      0  
143495  0.690382 -0.239318  0.051140  0.264748  0.124266    4.48      0  
154204  0.621209 -0.633905  0.416882 -0.045961 -0.038617   14.95      0  
141532 -0.272232  0.311050 -0.643594  0.120797  0.017599    1.00      0  
123720  0.030671 -0.897738  0.267859 -0.112916  0.156756   31.61      0  

[5 rows x 31 columns]

In [17]:
creditCard_new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [18]:
creditCard_new_df['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [19]:
creditCard_new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      95569.914634  0.028955 -0.039952  0.034604 -0.080581  0.077919   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.030335  0.013637  0.051963 -0.026660  ... -0.016369 -0.015112   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.023819 -0.024922  0.049736  0.025409  0.017194 -0.017906  0.009027   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       84.112683  
1      122.211321  

[2 rows x 30 columns]

In [20]:
# Splitting the new dataset/dataframe into input feature and target feature
x = creditCard_new_df.drop(columns = 'Class', axis = 1)
y = creditCard_new_df['Class']

In [21]:
print(x)

            Time        V1        V2        V3        V4        V5        V6  \
100818   67635.0  1.205203 -1.273859  1.688107 -0.269991 -2.002077  0.748282   
143495   85391.0 -0.438080  1.047487  1.698626  0.046575 -0.242688 -1.083626   
154204  100912.0  1.970590 -0.490162  0.070065  0.449218 -0.774299  0.054676   
141532   84378.0  1.116321 -0.521017  0.870835 -0.936391 -0.809979  0.433784   
123720   77036.0 -0.672101  0.605524  1.907994 -0.239642 -0.403889  0.266418   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [22]:
print(y)

100818    0
143495    0
154204    0
141532    0
123720    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [23]:
# Splitting the new dataset/df into training and testing data
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, stratify = y, random_state = 2)

In [24]:
print(x.shape, x_train.shape, x_test.shape)

(984, 30) (787, 30) (197, 30)


## Train the LR model using the above df

In [25]:
# Define the LR model
model = LogisticRegression()

In [26]:
# Train the LR model
model.fit(x_train, y_train)

LogisticRegression()

## Evaluate the model

In [27]:
# Accuracy score for the training data
x_train_prediction = model.predict(x_train)
training_data_accuracy = accuracy_score(x_train_prediction, y_train)
print("Accuracy of the training data: ", training_data_accuracy)

Accuracy of the trainig data:  0.9504447268106735


In [28]:
# Accuracy score for test data
x_test_prediction = model.predict(x_test)
test_data_accuracy = accuracy_score(x_test_prediction, y_test)
print("Accuracy of the test data:", test_data_accuracy)

Accuracy of the test data: 0.934010152284264


In [29]:
# 
joblib.dump(model, 'best_model.joblib')

['best_model.joblib']